### PR TITLE
Add a two amplitudes model

### PR DESCRIPTION
**Pull request by martin-ueding**
Model with different amplitude for forward and backwards part. See [Two_Amplitudes_Model.pdf](https://github.com/HISKP-LQCD/hadron/files/4408588/Two_Amplitudes_Model.pdf)

### CODE REVIEW

**Comment by kostrzewa** [comment]
Why two amplitudes?

**Comment by martin-ueding** [comment]
Carsten asked me how this could be done in an email, hinting at thermal pollution. I don't know in which cases this would be applicable, though.

**Comment by urbach** [comment]
For the hankel method in the presence of a backpropagating state with
the same mass a model like

A exp(-Mt) + Bexp(-M(T-t))

is needed...

File: R/new_matrixfit.R [hunk 1]
```diff
@@ -488,6 +520,9 @@ new_matrixfit <- function(cf,
   
   if (model == 'single') {
     model_object <- SingleModel$new(cf$Time, parlist, sym.vec, neg.vec, mSize)
+  } else if (model == 'two_amplitudes') {
```
Comment (urbach) [inline R/new_matrixfit.R]: doesn't this need to be mentioned in the roxygen docu as well?

File: R/new_matrixfit.R [hunk 2]
```diff
@@ -124,6 +124,38 @@ SingleModel <- R6::R6Class(
...
+  inherit = MatrixModel,
+  public = list(
+    initialize = function (time_extent, parlist, sym_vec, neg_vec, m_size) {
+      super$initialize(time_extent, parlist, sym_vec, neg_vec, m_size)
+    },
+    prediction = function (par, x, ...) {
+      sign_vec <- make_sign_vec(self$sym_vec, length(x), self$m_size)
+      ov_sign_vec <- make_ov_sign_vec(self$neg_vec, length(x), self$m_size)
+
+      ov_sign_vec * 0.5 * (par[2]^2 * exp(-par[1] * x) + sign_vec * par[3]^2 * exp(-par[1] * (self$time_extent - x)))
```
Comment (urbach) [inline R/new_matrixfit.R]: its important to document this factor 1/2 here...

File: R/new_matrixfit.R (hunk 1 above)
Comment (martin-ueding) [reply R/new_matrixfit.R]: Yes, it should be mentioned in the documentation. I have added formulas for most models.

Comment (martin-ueding) [reply R/new_matrixfit.R]: The weighted model is way more complicated:

```r
pred <- ov_sign_vec * (a0*(-((exp(2*e0*time) + exp(e0*(2*t + time))*sign)*Power(w,2*t)) - sign*(exp(2*e0*time) + exp(e0*(2*t + time))*sign)*Power(w,2*deltat + time) + (exp(e0*(deltat + 2*time)) + exp(e0*(-deltat + 2*t + time))*sign)*Power(w,deltat)* (Power(w,2*t) + sign*Power(w,time)))) / (exp(e0*(t + 2*time))*Power(w,deltat)*(Power(w,2*t) + sign*Power(w,time)))
```

I am not sure whether one could actually read it. I've derived it with Mathematica and just copied it out of there.

**Comment by urbach** [comment]
sorry, can we have it even a bit more general, by replacing

    exp(-E(T-t))  by  exp(E t)

? I can also do it myself...

**Comment by martin-ueding** [comment]
That should not be a big change, I'll do it.

**Comment by urbach** [comment]
why do all the checks fail here?

**Comment by martin-ueding** [comment]
Apparently there has been a change in R versions or the Travis environment and also the `master` fails. I try to fix that.

File: R/new_matrixfit.R [hunk 3]
```diff
@@ -124,6 +124,38 @@ SingleModel <- R6::R6Class(
...
+      ov_sign_vec <- make_ov_sign_vec(self$neg_vec, length(x), self$m_size)
+
+      ov_sign_vec * 0.5 * (par[2]^2 * exp(-par[1] * x) + sign_vec * par[3]^2 * exp(par[1] x))
+    },
+    prediction_jacobian = function (par, x, ...) {
+      sign_vec <- make_sign_vec(self$sym_vec, length(x), self$m_size)
+      ov_sign_vec <- make_ov_sign_vec(self$neg_vec, length(x), self$m_size)
+
+      res <- matrix(NA, nrow = length(x), ncol = length(par))
+      res[, 1] <- ov_sign_vec * 0.5 * (par[2]^2 * exp(-par[1] * x) * (-x) + sign_vec * par[3]^2 * exp(par[1] * x) * x
```
Comment (urbach) [inline R/new_matrixfit.R:145]: a closing `)` missing?

File: R/new_matrixfit.R [hunk 4]
```diff
@@ -124,6 +124,38 @@ SingleModel <- R6::R6Class(
...
+  inherit = MatrixModel,
+  public = list(
+    initialize = function (time_extent, parlist, sym_vec, neg_vec, m_size) {
+      super$initialize(time_extent, parlist, sym_vec, neg_vec, m_size)
+    },
+    prediction = function (par, x, ...) {
+      sign_vec <- make_sign_vec(self$sym_vec, length(x), self$m_size)
+      ov_sign_vec <- make_ov_sign_vec(self$neg_vec, length(x), self$m_size)
+
+      ov_sign_vec * 0.5 * (par[2]^2 * exp(-par[1] * x) + sign_vec * par[3]^2 * exp(par[1] x))
```
Comment (urbach) [inline R/new_matrixfit.R:138]: a `*` missing before the last `x`

Comment (martin-ueding) [reply R/new_matrixfit.R]: Okay, fixed that.

**Pull request merged**